### PR TITLE
support resign leader.

### PR DESCRIPTION
**Pull request by disksing**
HTTP API:
```
$ curl -X POST http://127.0.0.1:2379/pd/api/v1/leader/resign
$ curl -X POST http://127.0.0.1:2379/pd/api/v1/leader/transfer/pd2
```
pd-ctl:
```
> member leader resign
> member leader transfer pd2
```

### CODE REVIEW

**Comment by siddontang** [comment]
PTAL @huachaohuang

File: server/api/member_test.go [hunk 1]
```diff
@@ -193,3 +193,46 @@ func (s *testMemberAPISuite) TestMemberLeader(c *C) {
...
+	addrs := make(map[uint64]string)
+	for i := range cfgs {
+		addrs[svrs[i].ID()] = cfgs[i].ClientUrls
+	}
+
+	leader1, err := svrs[0].GetLeader()
+	c.Assert(err, IsNil)
+
+	s.post(c, addrs[leader1.GetMemberId()]+apiPrefix+"/api/v1/leader/resign")
+	leader2 := s.waitLeaderChange(c, svrs[0], leader1)
```
Comment (huachaohuang) [inline server/api/member_test.go]: Should check leader2 != leader1?

Comment (disksing) [inline server/api/member_test.go]: No. It's checked inside `waitLeaderChange`.

**Comment by siddontang** [comment]
PTAL @nolouch

File: server/api/member.go [hunk 2]
```diff
@@ -135,3 +135,23 @@ func (h *leaderHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
+	if err != nil {
+		h.rd.JSON(w, http.StatusInternalServerError, err.Error())
+		return
+	}
+
+	h.rd.JSON(w, http.StatusOK, nil)
+}
+
+func (h *leaderHandler) Transfer(w http.ResponseWriter, r *http.Request) {
+	err := h.svr.ResignLeader(mux.Vars(r)["next_leader"])
```
Comment (siddontang) [inline server/api/member.go]: here is not next leader but new leader?

Comment (disksing) [inline server/api/member.go]: I think no big difference to use _next_ or _new_.

File: server/leader.go [hunk 3]
```diff
@@ -259,7 +302,51 @@ func (s *Server) watchLeader() {
...
+			leaderIDs = append(leaderIDs, strconv.FormatUint(member.ID, 10))
+		}
+	}
+	nextLeaderValue := strings.Join(leaderIDs, ",")
+
+	// Save expect leader(s) to etcd.
+	lease := clientv3.NewLease(s.client)
+	defer lease.Close()
+	ctx, cancel := context.WithTimeout(s.client.Ctx(), requestTimeout)
+	leaseRes, err := lease.Grant(ctx, int64(nextLeaderTTL))
```
Comment (siddontang) [inline server/leader.go]: do we need to use config leader lease instead of `nextLeaderTTL`?

Comment (disksing) [inline server/leader.go]: We can make them the same. But this TTL doesn't have direct relationship with leader lease, it's ok to be different, no matter shorter or longer.

File: server/leader.go [hunk 4]
```diff
@@ -259,7 +302,51 @@ func (s *Server) watchLeader() {
 	}
 }
 
-func (s *Server) resignLeader() error {
+// ResignLeader resigns current PD's leadership. If nextLeader is empty, all
+// other pd-servers can campaign.
+func (s *Server) ResignLeader(nextLeader string) error {
+	// Determine next leaders.
+	var leaderIDs []string
```
Comment (siddontang) [inline server/leader.go]: add log here

File: server/leader.go [hunk 5]
```diff
@@ -259,7 +302,51 @@ func (s *Server) watchLeader() {
...
+		Commit()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if !resp.Succeeded {
+		return errors.New("save next leader failed, maybe lost leadership")
+	}
+
+	// Resign leader.
+	select {
```
Comment (siddontang) [inline server/leader.go]: log next leader here

**Pull request merged**